### PR TITLE
chore: simplify IndexScalar

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

_Briefly describe what this PR aims to solve. Include background context that will help reviewers understand the purpose of the PR._

- Fixes #[Link the issue here]

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14432)

### CODE REVIEW

**Comment by andylokandy** [comment]
After  https://github.com/datafuselabs/databend/pull/14368, `AppendGenerator` still generates ColumnStatistics with Scalar::Map. @sundy-li @dantengsky 

```sql
> INSERT INTO TABLE `05_0028_at_t0` (a, m) values(1.0, {2:'mn'})
```

```rust
[src/query/storages/fuse/src/operations/common/processors/sink_commit.rs:253] &snapshot = TableSnapshot {
    format_version: 4,
    snapshot_id: 888200a8-5557-4922-acb9-3c6252af6da5,
    timestamp: Some(
        2024-01-23T11:15:19.552805Z,
    ),
    prev_snapshot_id: Some(
        (
            233a25b4-d5b2-4bed-8737-e98469073a74,
            4,
        ),
    ),
    schema: TableSchema {
        fields: [
            TableField {
                name: "a",
                default_expr: Some(
                    "101",
                ),
                data_type: Number(
                    Float32,
                ),
                column_id: 9,
                computed_expr: None,
            },
            TableField {
                name: "b",
                default_expr: Some(
                    "(202, 13)",
                ),
                data_type: Tuple {
                    fields_name: [
                        "1",
                        "2",
                    ],
                    fields_type: [
                        Nullable(
                            Number(
                                Float64,
                            ),
                        ),
                        Nullable(
                            Number(
                                Int32,
                            ),
                        ),
                    ],
                },
                column_id: 10,
                computed_expr: None,
            },
            TableField {
                name: "t",
                default_expr: Some(
                    "((1, 2), (3, 4))",
                ),
                data_type: Tuple {
                    fields_name: [
                        "a",
                        "b",
                    ],
                    fields_type: [
                        Nullable(
                            Tuple {
                                fields_name: [
                                    "m",
                                    "n",
                                ],
                                fields_type: [
                                    Nullable(
                                        Number(
                                            Int64,
                                        ),
                                    ),
                                    Nullable(
                                        Number(
                                            Float64,
                                        ),
                                    ),
                                ],
                            },
                        ),
                        Nullable(
                            Tuple {
                                fields_name: [
                                    "x",
                                    "y",
                                ],
                                fields_type: [
                                    Nullable(
                                        Number(
                                            Int64,
                                        ),
                                    ),
                                    Nullable(
                                        Number(
                                            Int64,
                                        ),
                                    ),
                                ],
                            },
                        ),
                    ],
                },
                column_id: 13,
                computed_expr: None,
            },
            TableField {
                name: "m",
                default_expr: Some(
                    "{1:'ab'}",
                ),
                data_type: Map(
                    Tuple {
                        fields_name: [
                            "key",
                            "value",
                        ],
                        fields_type: [
                            Number(
                                Int64,
                            ),
                            Nullable(
                                String,
                            ),
                        ],
                    },
                ),
                column_id: 17,
                computed_expr: None,
            },
        ],
        metadata: {},
        next_column_id: 19,
    },
    summary: Statistics {
        row_count: 5,
        block_count: 5,
        perfect_block_count: 0,
        uncompressed_byte_size: 430,
        compressed_byte_size: 4845,
        index_size: 2255,
        col_stats: {
            17: ColumnStatistics {
                min: Number(
                    2_i64,
                ),
                max: Map(
                    Tuple(
                        [
                            Int64([1]),
                            NullableColumn { column: StringColumn { data: 0x6162, offsets: [0, 2] }, validity: [0b_______1] },
                        ],
                    ),
                ),
                null_count: 0,
                in_memory_size: 8,
                distinct_of_values: None,
            },
            18: ColumnStatistics {
                min: String(
                    "mn",
                ),
                max: String(
                    "mn",
                ),
                null_count: 0,
                in_memory_size: 19,
                distinct_of_values: None,
            },
            12: ColumnStatistics {
                min: Number(
                    0_i32,
                ),
                max: Number(
                    202_i32,
                ),
                null_count: 0,
                in_memory_size: 4,
                distinct_of_values: None,
            },
            5: ColumnStatistics {
                min: Date(
                    18262,
                ),
                max: Date(
                    18262,
                ),
                null_count: 0,
                in_memory_size: 5,
                distinct_of_values: None,
            },
            3: ColumnStatistics {
                min: Number(
                    12.34_f64,
                ),
                max: Number(
                    12.34_f64,
                ),
                null_count: 0,
                in_memory_size: 9,
                distinct_of_values: None,
            },
            7: ColumnStatistics {
                min: Number(
                    2_u8,
                ),
                max: Number(
                    101_u8,
                ),
                null_count: 0,
                in_memory_size: 9,
                distinct_of_values: None,
            },
            4: ColumnStatistics {
                min: String(
                    "abc",
                ),
                max: String(
                    "abc",
                ),
                null_count: 0,
                in_memory_size: 20,
                distinct_of_values: None,
            },
            15: ColumnStatistics {
                min: Number(
                    3_i64,
                ),
                max: Number(
                    5_i64,
                ),
                null_count: 0,
                in_memory_size: 18,
                distinct_of_values: None,
            },
            11: ColumnStatistics {
                min: Number(
                    2_i32,
                ),
                max: Number(
                    13_i32,
                ),
                null_count: 0,
                in_memory_size: 15,
                distinct_of_values: None,
            },
            13: ColumnStatistics {
                min: Number(
                    1_i64,
                ),
                max: Number(
                    3_i64,
                ),
                null_count: 0,
                in_memory_size: 18,
                distinct_of_values: None,
            },
            8: ColumnStatistics {
                min: Number(
                    4_f64,
                ),
                max: Number(
                    4_f64,
                ),
                null_count: 0,
                in_memory_size: 8,
                distinct_of_values: None,
            },
            10: ColumnStatistics {
                min: Number(
                    111_f64,
                ),
                max: Number(
                    202_f64,
                ),
                null_count: 0,
                in_memory_size: 27,
                distinct_of_values: None,
            },
            6: ColumnStatistics {
                min: Timestamp(
                    1577836800000000,
                ),
                max: Timestamp(
                    1577836800000000,
                ),
                null_count: 0,
                in_memory_size: 9,
                distinct_of_values: None,
            },
            2: ColumnStatistics {
                min: Number(
                    100_i64,
                ),
                max: Number(
                    100_i64,
                ),
                null_count: 0,
                in_memory_size: 9,
                distinct_of_values: None,
            },
            9: ColumnStatistics {
                min: Number(
                    1_f32,
                ),
                max: Number(
                    102_f32,
                ),
                null_count: 0,
                in_memory_size: 16,
                distinct_of_values: None,
            },
            14: ColumnStatistics {
                min: Number(
                    2_f64,
                ),
                max: Number(
                    4_f64,
                ),
                null_count: 0,
                in_memory_size: 18,
                distinct_of_values: None,
            },
            0: ColumnStatistics {
                min: Number(
                    1_i32,
                ),
                max: Number(
                    1_i32,
                ),
                null_count: 0,
                in_memory_size: 4,
                distinct_of_values: None,
            },
            16: ColumnStatistics {
                min: Number(
                    4_i64,
                ),
                max: Number(
                    6_i64,
                ),
                null_count: 0,
                in_memory_size: 18,
                distinct_of_values: None,
            },
        },
        cluster_stats: None,
    },
    segments: [
        (
            "10070/10077/_sg/728e421698cb4a84afabdcee3d1ad8fe_v4.mpk",
            4,
        ),
        (
            "10070/10077/_sg/eb9272c7591045a7ab8a9e4c24de4fa5_v4.mpk",
            4,
        ),
        (
            "10070/10077/_sg/cec9830474204a5eb8e1970707b4e4bd_v4.mpk",
            4,
        ),
        (
            "10070/10077/_sg/9f8db12c6d234408a9dc9f5498aa8b85_v4.mpk",
            4,
        ),
        (
            "10070/10077/_sg/489a8e3693ec47bbb4ade8a9363929a5_v4.mpk",
            4,
        ),
    ],
    cluster_key_meta: None,
    table_statistics_location: None,
}
```

File: src/query/storages/common/table_meta/src/meta/v2/statistics.rs [hunk 1]
```diff
@@ -69,20 +89,20 @@ impl ColumnStatistics {
...
+            max,
             null_count,
             in_memory_size,
             distinct_of_values,
         }
     }
 
     pub fn min(&self) -> Scalar {
-        self.min.clone().into()
+        self.min.clone()
```
Comment (sundy-li) [inline src/query/storages/common/table_meta/src/meta/v2/statistics.rs:101]: If so, we just need to return the reference now.

File: src/query/storages/common/table_meta/src/meta/v2/statistics.rs [hunk 2]
```diff
@@ -69,20 +89,20 @@ impl ColumnStatistics {
         distinct_of_values: Option<u64>,
     ) -> Self {
         Self {
```
Comment (sundy-li) [inline src/query/storages/common/table_meta/src/meta/v2/statistics.rs:91]: Maybe we could add debug_assert to assume that:

min, max are both same enum types or at least one of them is null.

**Comment by dantengsky** [comment]
> After #14368, `AppendGenerator` still generates ColumnStatistics with Scalar::Map. @sundy-li @dantengsky.
> 
> PS: the min max types also seems different, is it a bug?
> ....

@andylokandy 

failed to reproduce the problem:

~~~
mysql> desc t_05_0028_at_t0;
+-------+--------------------+------+---------+-------+
| Field | Type               | Null | Default | Extra |
+-------+--------------------+------+---------+-------+
| a     | FLOAT              | YES  | NULL    |       |
| m     | MAP(INT32, STRING) | YES  | NULL    |       |
+-------+--------------------+------+---------+-------+
2 rows in set (0.04 sec)
Read 0 rows, 0.00 B in 0.013 sec., 0 rows/sec., 0.00 B/sec.

mysql> delete from t_05_0028_at_t0;
Query OK, 2 rows affected (0.66 sec)

mysql> insert into t_05_0028_at_t0 (a, m) values(1.0, {2:'a'}), (2.0, {3: 'b'});
Query OK, 2 rows affected (0.10 sec)

mysql> select version();
+----------------------------------------------------------------------------------------+
| version()                                                                              |
+----------------------------------------------------------------------------------------+
| 8.0.26-v1.2.312-nightly-c6e35735d9(rust-1.75.0-nightly-2024-01-23T11:46:20.471802059Z) |
+----------------------------------------------------------------------------------------+
1 row in set (0.02 sec)
Read 1 rows, 1.00 B in 0.004 sec., 254.37 rows/sec., 254.37 B/sec.
~~~



The corresponding snapshot is :



~~~
snapshot is TableSnapshot {
    format_version: 4,
    snapshot_id: 91e7abdd-9bfd-4021-ae1c-623a85cfa861,
    timestamp: Some(
        2024-01-23T12:05:49.864315Z,
    ),
    prev_snapshot_id: Some(
        (
            0392ff5c-7ca5-4208-8da2-8dec78591fbc,
            4,
        ),
    ),
    schema: TableSchema {
        fields: [
            TableField {
                name: "a",
                default_expr: None,
                data_type: Nullable(
                    Number(
                        Float32,
                    ),
                ),
                column_id: 0,
                computed_expr: None,
            },
            TableField {
                name: "m",
                default_expr: None,
                data_type: Nullable(
                    Map(
                        Tuple {
                            fields_name: [
                                "key",
                                "value",
                            ],
                            fields_type: [
                                Number(
                                    Int32,
                                ),
                                Nullable(
                                    String,
                                ),
                            ],
                        },
                    ),
                ),
                column_id: 1,
                computed_expr: None,
            },
        ],
        metadata: {},
        next_column_id: 3,
    },
    summary: Statistics {
        row_count: 2,
        block_count: 1,
        perfect_block_count: 0,
        uncompressed_byte_size: 69,
        compressed_byte_size: 614,
        index_size: 337,
        col_stats: {
            0: ColumnStatistics {
                min: Number(
                    1_f32,
                ),
                max: Number(
                    2_f32,
                ),
                null_count: 0,
                in_memory_size: 9,
                distinct_of_values: None,
            },
            2: ColumnStatistics {
                min: String(
                    [
                        97,
                    ],
                ),
                max: String(
                    [
                        98,
                    ],
                ),
                null_count: 0,
                in_memory_size: 27,
                distinct_of_values: None,
            },
            1: ColumnStatistics {
                min: Number(
                    2_i32,
                ),
                max: Number(
                    3_i32,
                ),
                null_count: 0,
                in_memory_size: 8,
                distinct_of_values: None,
            },
        },
        cluster_stats: None,
    },
    segments: [
        (
            "1/10309/_sg/a51af09d120a4c489054307573706065_v4.mpk",
            4,
        ),
    ],
    cluster_key_meta: None,
    table_statistics_location: None,
}
~~~

**Comment by andylokandy** [comment]
@sundy-li PTAL

**Review by sundy-li — APPROVED** [review]

**Pull request merged**